src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,294 @@
+                        nameRef.WithoutTrivia());
+                }
+                else
+                {
+                    var aliasQualifier = GetAliasQualifierOpt(old);
+                    var qualifiedNamespaceName = CreateNameSyntax(newNamespaceParts, aliasQualifier, newNamespaceParts.Length - 1);
+                    @new = SyntaxFactory.QualifiedName(qualifiedNamespaceName, nameRef.WithoutTrivia());
+                }
+
+                // We might lose some trivia associated with children of `outerMostNode`.  
+                @new = @new.WithTriviaFrom(old);
+            }
+            return true;
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier?.EscapeIdentifier();
+
+        protected override async Task<(bool, NamespaceDeclarationSyntax)> ShouldPositionTriggerRefactoringAsync(
+            Document document, 
+            int position, 
+            CancellationToken cancellationToken)
+        {
+            var compilationUnit = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false) as CompilationUnitSyntax;            
+            var namespaceDecls = compilationUnit.DescendantNodes().OfType<NamespaceDeclarationSyntax>().ToImmutableArray();
+
+            // Here's conditions that trigger the refactoring (all have to be true in each scenario):
+            // 
+            // - There's only one namespace declaration in the document and all types are declared in it:
+            //    1. No nested namespace declaration (even it's empty).
+            //    2. The cursor is on the name of the namespace declaration.
+            //    3. The name of the namespace is valid (i.e. no errors).
+            //    4. No partial type declared in the namespace. Otherwise its multiple declaration will
+            //       end up in different namespace.
+            //
+            // - There's no namespace declaration and all types in the document are declared in global namespace:
+            //    1. The cursor is on the name of first declared type.
+            //    2. No partial type declared in the document. Otherwise its multiple declaration will
+            //       end up in different namespace.
+
+            if (namespaceDecls.Length == 1 && compilationUnit.Members.Count == 1)
+            {
+                var namespaceDeclaration = namespaceDecls.Single();
+                Debug.Assert((object)namespaceDeclaration == (object)compilationUnit.Members.Single());
+
+                var shouldTrigger = namespaceDeclaration.Name.Span.IntersectsWith(position) 
+                    && namespaceDeclaration.Name.GetDiagnostics().All(diag => diag.DefaultSeverity != DiagnosticSeverity.Error);
+
+                shouldTrigger = shouldTrigger && 
+                    !(await ContainsPartialTypeWithMultipleDeclarationsAsync(document, namespaceDeclaration, cancellationToken)
+                    .ConfigureAwait(false));
```

Comment:
fwiw, this is pretty unpleasant to read.  nested expressions with awaits just get unweildy.